### PR TITLE
DefaultSenderPlugin - allows to specify default sender that is used when a message with no sender is specified

### PR DESCRIPTION
**Pull request by adamziel**
DefaultSenderPlugin - allows to specify default sender that is used when a message with no sender is sent

### CODE REVIEW

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * @author     Adam Zielinski
+ */
+class Swift_Plugins_DefaultSenderPlugin implements Swift_Events_SendListener
+{
+
+    /**
+     * The default sender email.
+     *
+     * @var String
```
Comment (xabbuh) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: `String` should be lowercased.

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    private $_defaultSenderEmail;
```
Comment (xabbuh) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Afaik, Swiftmailer doesn't use underscore-prefixed names for private properties.

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * The default sender name.
+     *
+     * @var String
+     */
+    private $_defaultSenderName;
+
+    /**
+     * List if IDs of handled messages
+     *
+     * @var array
```
Comment (xabbuh) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: You could use `string[]` here.

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    private $_handledMessageIds = array();
+
+    /**
+     * Create a new ImpersonatePlugin to impersonate $sender.
+     *
+     * @param string $sender address
```
Comment (xabbuh) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Missing documentation for the `$defaultSenderName` argument.

**Comment by AZielinski** [comment]
I updated this pull request to reflect the changes suggested by @xabbuh

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * List if IDs of handled messages
+     *
+     * @var string[]
+     */
+    private $handledMessageIds = array();
+
+    /**
+     * Create a new ImpersonatePlugin to impersonate $sender.
```
Comment (stof) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: wrong phpdoc

File: tests/unit/Swift/Plugins/DefaultSenderPluginTest.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        $this->assertEquals($message->getFrom(), array('john@example.com' => 'John Doe'));
+
+        $plugin->sendPerformed($evt);
+
+        $this->assertEquals($message->getFrom(), array());
+    }
+
+    // -- Creation Methods
+
+    private function _createSendEvent(Swift_Mime_Message $message)
```
Comment (stof) [inline tests/unit/Swift/Plugins/DefaultSenderPluginTest.php]: the leading underscore should be removed

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php (hunk 5 above)
Comment (AZielinski) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: What do you mean exactly? I'd love to make it correct, please just point me in the right direction

Comment (stof) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Well, read the line again and tell me whether it actually describes the method. Copy-pastes are bad

File: tests/unit/Swift/Plugins/DefaultSenderPluginTest.php (hunk 6 above)
Comment (AZielinski) [inline tests/unit/Swift/Plugins/DefaultSenderPluginTest.php]: Fixed. I used this file as a reference:
https://github.com/swiftmailer/swiftmailer/blob/master/tests/unit/Swift/Plugins/DecoratorPluginTest.php
which could mean there is some work to do in regard to code style in other files too.

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php (hunk 5 above)
Comment (AZielinski) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Just improved it

Comment (stof) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: the phpdoc is still describing the wrong method

Comment (AZielinski) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: oops in "Conversation" tab it looked like you commented on phpdoc of properties so I improved documentation of properties and missed the constructor comment. Now it's fixed

**Comment by sstok** [comment]
Looks good to me 👍

**Comment by mekras** [comment]
Any plans to merge this?

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
+<?php
+/*
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Missing empty line after `<?php`

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * This file is part of SwiftMailer.
+ * (c) 2009 Fabien Potencier
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: should be 2016

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Sets the sender of a message if none is set
+ *
+ * @author     Adam Zielinski
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Too many whitespace after `@author`

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+/*
+ * This file is part of SwiftMailer.
+ * (c) 2009 Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Sets the sender of a message if none is set
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: Missing dot at the end

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ */
+class Swift_Plugins_DefaultSenderPlugin implements Swift_Events_SendListener
+{
+
+    /**
+     * Email of sender to use in all messages that are sent
+     * without any sender information.
+     *
+     * @var string
+     */
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: I would remove the block and add the description in the `@param` of the constructor argument (same for `$defaultSenderName` below).

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    private $defaultSenderName;
+
+    /**
+     * List if IDs of messages which got sender information set
+     * by this plugin.
+     *
+     * @var string[]
+     */
+    private $handledMessageIds = array();
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: This is purely internal, so I would remove the docbloc.

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Create a new DefaultSenderPlugin to use a $defaultSenderEmail and $defaultSenderName
+     * for all messages that are sent without any sender information.
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: This can be removed

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        $this->defaultSenderEmail = $defaultSenderEmail;
+        $this->defaultSenderName = $defaultSenderName;
+    }
+
+    /**
+     * Invoked immediately before the Message is sent.
+     *
+     * @param Swift_Events_SendEvent $evt
+     */
+    public function beforeSendPerformed(Swift_Events_SendEvent $evt)
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: should be

```
/**
 * @{inheritdoc}
 */
```

instead

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            $this->handledMessageIds[$message->getId()] = true;
+        }
+    }
+
+    /**
+     * Invoked immediately after the Message is sent.
+     *
+     * @param Swift_Events_SendEvent $evt
+     */
+    public function sendPerformed(Swift_Events_SendEvent $evt)
```
Comment (fabpot) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: same as above

File: tests/unit/Swift/Plugins/DefaultSenderPluginTest.php [hunk 16]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+class Swift_Plugins_DefaultSenderPluginTest extends \PHPUnit_Framework_TestCase
+{
+
```
Comment (fabpot) [inline tests/unit/Swift/Plugins/DefaultSenderPluginTest.php]: extra empty line should be removed

File: tests/unit/Swift/Plugins/DefaultSenderPluginTest.php [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        $plugin->beforeSendPerformed($evt);
+
+        $this->assertEquals($message->getFrom(), array('john@example.com' => 'John Doe'));
+
+        $plugin->sendPerformed($evt);
+
+        $this->assertEquals($message->getFrom(), array());
+    }
+
+    // -- Creation Methods
```
Comment (fabpot) [inline tests/unit/Swift/Plugins/DefaultSenderPluginTest.php]: Can be removed

**Review by fabpot — APPROVED** [review]
Looks like a good plugin to add. Can you also add some documentation for that new feature and add an entry in the changelog? Thanks.

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Invoked immediately before the Message is sent.
+     *
+     * @param Swift_Events_SendEvent $evt
+     */
+    public function beforeSendPerformed(Swift_Events_SendEvent $evt)
+    {
+        $message = $evt->getMessage();
+        $headers = $message->getHeaders();
```
Comment (pierredup) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: `$headers` is not used in this method

**Comment by ScreamZ** [comment]
@adamziek Can you make changes in order to merge this PR pls.

Thanks

File: lib/classes/Swift/Plugins/DefaultSenderPlugin.php (hunk 14 above)
Comment (FinesseRus) [inline lib/classes/Swift/Plugins/DefaultSenderPlugin.php]: The correct keyword is {@inheritDoc}

**Comment by FinesseRus** [comment]
If you need to set a default sender you can use this plugin: [Swift Mailer Defaults Plugin](https://github.com/FinesseRus/SwiftMailerDefaultsPlugin)

**Comment by adamziel** [comment]
@ScreamZ I updated formatting in this PR

**Comment by fabpot** [comment]
Thanks for the PR. As announced some time ago, Swiftmailer is not maintained anymore.

[Symfony Mailer](https://symfony.com/doc/current/mailer.html) is the next evolution of Swiftmailer.
It provides the same features with support for modern PHP code and support for third-party providers.

**Pull request closed without merging**